design.md

```diff
@@ -0,0 +1,100 @@
+# V1
+
+The current code base I will call "v1" (referred to below as the current
+version as well).
+
+The current version is working fairly well in production (overall data
+transfer for tc artifacts is now about $2-8~ dollars a day when
+transferring in aws). In that sense things are great... There are a
+number of other issues we have though:
+
+
+  - The current design is not very flexible (one source -> dest)
+
+      This problem is very very common... You want a canonical source
+      for s3 where you can easily view what's going on but you don't want to
+      incur thousands of dollars of data transfer costs at scale.
+
+      For us to handle more common cases we need a real api.
+
+  - The current design is not trivially scalable:
+
+      We must deploy in _every_ aws region we care about and reconfigure
+      individual callers to be smart enough to pick the right regions that
+      happen to be configured....
+
+      In the worst case a single node goes bad (we have seen this
+      already!) and the node manually requires cycling.
+
+
+## V2
+
+V2 is a series of major changes to how we do both the uploading and
+proxing of things. To start we break up the service into two parts.
+
+  1. The uploader
+  2. The proxy
+
+The uploader will be fairly dumb and only care about uploading stuff
+it's told to upload. It will exist in one region (and eventually) will
+autoscale.
+
+The proxy will be deployed (most likely on heroku) in a fashion that
+enables us to encode all the region detection logic and will be
+responsible for containing the configs of what paths map to what sources
+and metrics for end-to-end experience.
+
+### Proxy
+
+TODO: More details / maybe a graph.
+
+  - Detect AWS Region if possiblee
```

Comment:
yeah, we use [ip-ranges.json](https://ip-ranges.amazonaws.com/ip-ranges.json), we already do this:
https://github.com/taskcluster/taskcluster-queue/blob/e4eec406982dfcbef93157087ab588187448d2a2/queue/ec2regionresolver.js